apps/metagpt/gpt.py

```diff
@@ -0,0 +1,109 @@
+
+    message_content: str
+    cost: float
+    _message_contnent_dix: Dict[str, Any] | None = None
+
+    @property
+    def message_content_as_dict(self: Self) -> Dict[str, Any]:
+        """Message content from GPT as dictionary."""
+        if self._message_contnent_dix is None:
+            self._message_contnent_dix = yaml.safe_load(self.message_content)
+        else:
+            raise ValueError("`message_content` is empty!")
+        return self._message_contnent_dix
+
+
+@dataclass
+class GPTQuery:
+    """Fields for GPT query."""
+
+    messages: List[Dict[str, str]]
+    temperature: float = 0
+
+    @property
+    def estimated_cost(self: Self) -> float:
+        """
+        Calculate the cost of using GPT based on the number of characters of the message content.
+
+        This function estimates the cost of using GPT by converting the number of characters into tokens,
+        rounding up to the nearest thousand tokens, and then multiplying by the rate per thousand tokens.
+
+        The cost is given in USD.
+
+        Returns:
+            float: The estimated cost of using GPT for the given number of characters.
+        """
+        char_count = len(self.messages[0]["content"])  # type: ignore
+        tokens = char_count / 4  # Average size of a token is 4 characters
+        tokens_rounded_up = -(-tokens // 1000) * 1000  # Round up to the nearest 1000 tokens
+        estimated_cost = (tokens_rounded_up / 1000) * RATE_PER_1000_TOKENS
+        return estimated_cost
+
+    def to_dict(self: Self) -> Dict[str, Any]:
+        """Class as dictionary."""
+        return {k: v for k, v in asdict(self).items()}
+
+
+class OpenAIWrapper(OpenAI):
+    """Wrapper for OpenAI API."""
+
+    def __init__(self: Self, **kwargs) -> None:
+        """Initialize OpenAI API wrapper."""
+        super().__init__(**kwargs)
+        self.model = kwargs.get("model", GPT_MODEL)
+
+    def query_gpt(self: Self, query: GPTQuery | None = None, **kwargs) -> GPTResponse | None:
+        """Query Chat GPT to get message content from the chat completion."""
+        # Get chat completion
+        _ = kwargs.pop("model", None)
+        if query:
+            kwargs = {
+                "model": self.model,
+                **query.to_dict(),
+            }
+        else:
+            kwargs = {
+                "model": self.model,
+                **kwargs,
+            }
+        chat_completion = self.chat.completions.create(**kwargs)  # type: ignore
+        # Return value only if message content is not None
+        if isinstance(chat_completion, ChatCompletion) and isinstance(chat_completion.usage, CompletionUsage):
+            chat_completion_tokens = chat_completion.usage.total_tokens
+            cost = (chat_completion_tokens / 1000) * RATE_PER_1000_TOKENS
+            message_content = chat_completion.choices[0].message.content
+            # Log cost
+            log.info(f"Chat completion {self.model} cost: {cost}")
```

Comment:
we probably don't need to log the cost at each step but what we could do is implement a progress bar for garden/grapher step to show where we are in the loop of indicators+ metadata field updates

Reply:
I can easily make it work in the terminal but not sure how to do it in wizard afterwards...

Reply:
we can do this in a future PR